File: 2094-remove-stones-to-minimize-the-total/remove-stones-to-minimize-the-total.cpp
class Solution {
public:
    int minStoneSum(vector<int>& piles, int k) {
        vector<int> fre(10001, 0);
        // int m = piles[0];
        for(auto i: piles){
            fre[i]++;
        }
        for(int i = 10000; i>0 && k > 0; i--){
            while(fre[i] > 0 && k >0){
                fre[i]--;
                fre[(i+1)/2]++;
                k--;
            }
        }
        int sum = 0;
        for(int i = 0; i <= 10000; i++){
            sum += i*fre[i];
        }
        return sum;
    }
};

File: 2094-remove-stones-to-minimize-the-total/remove-stones-to-minimize-the-total.java
class Solution {
    public int minStoneSum(int[] piles, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b-a);
        for(int i = 0; i < piles.length; i++){
            pq.add(piles[i]);
        }

        while(k > 0){
            int m = pq.poll();
            m = (int)(m - Math.floor(m/2));
            pq.add(m);
            k--;
        }
        int sum = 0;
        while(!pq.isEmpty()){
            sum += pq.poll();
        }
        return sum;
    }
}